packages/bindings-factory/lib/Bindings.ts

```diff
@@ -138,13 +191,93 @@ export class Bindings implements RDF.Bindings {
       entries.push([ key, value ]);
     }
 
-    return new Bindings(this.dataFactory, Map(entries));
+    if (this.context) {
+      let mergedContext = this.context;
+      // Only merge if the other has a context
+      if ('context' in other && other.context) {
+        mergedContext = this.mergeContext(mergedContext, other.context);
+      }
+      return new Bindings(this.dataFactory, Map(entries), this.contextMergeHandlers, mergedContext);
+    }
+    const castOther = <Bindings> other;
+    return new Bindings(this.dataFactory, Map(entries), this.contextMergeHandlers, castOther.context);
+  }
+
+  private mergeContext(context: IActionContext, otherContext: IActionContext): IActionContext {
+    // Get Set of all keys present in either of the bindings
+    const keysContext = this.uniqueKeys([ ...context.keys(),
+      ...otherContext.keys() ]);
+
+    // Get Set of all keys present in both bindings
+    const keysBothContext = context.keys().filter(
+      element => otherContext.keys().some(({ name }) => element.name === name),
+    );
+    // Merge context based on supplied mergeHandlers
+    for (const key of keysContext) {
+      const keyString = key.name;
+      const occursInBoth = keysBothContext.some(x => x.name === key.name);
```

Comment:
If `keysBothContext` would be a set, we can very cheaply call `.includes()` on it.